File: examples/swift/Sources/FluentKit/Catalog.swift
// SPDX-FileCopyrightText: © 2026 Jeffrey C. Ollie <jeff@ocjtech.us>
// SPDX-License-Identifier: MIT

//
//  The translations an application ships, and the choice between them.
//
//  This is `examples/common/catalog.c` in Swift, and deliberately so: the two
//  GUI examples written in C share that file, and reading this one beside it
//  shows what changes when the language does and what does not. The
//  negotiation is the same three passes, for the same reason -- it is the
//  application's decision rather than the library's.
//

import Foundation

public struct Translation {
	public let tag: String
	/// The name this translation gives its own language: "Deutsch", not
	/// "German". It comes from the `language-name` message inside the
	/// translation, which is the only place that knows it.
	public let endonym: String
	let bundle: FluentBundle
}

public enum CatalogError: Error, CustomStringConvertible {
	case noBundle(String)
	case unreadable(String, underlying: Error)
	case malformed(String, problems: [String])
	case outOfMemory

	public var description: String {
		switch self {
		case .noBundle(let tag): "\(tag) is not a language tag"
		case .unreadable(let tag, let underlying): "\(tag).ftl: \(underlying.localizedDescription)"
		case .malformed(let tag, let problems): "\(tag).ftl: \(problems.joined(separator: "; "))"
		case .outOfMemory: "out of memory"
		}
	}
}

public final class Catalog {
	/// The first entry is the source locale and is the last resort, which is
	/// why its translation must be complete.
	public static let shipped = ["en-US", "de", "fr", "fi", "pl", "ja"]

	public private(set) var translations: [Translation] = []

	/// One list, cleared by the caller between screenfuls, so that a window
	/// can format a dozen messages and then ask once what went wrong.
	public let errors: FluentErrors

	public init(directory: URL) throws {
		guard let errors = FluentErrors() else { throw CatalogError.outOfMemory }
		self.errors = errors

		for tag in Self.shipped {
			guard let bundle = FluentBundle(locale: tag) else {
				throw CatalogError.noBundle(tag)
			}

			// Left on, unlike a command-line tool: these strings are
			// going into a window, where the isolation marks do their
			// job and are invisible.
			bundle.setUsesIsolating(true)

			let file = directory.appendingPathComponent("\(tag).ftl")
			let source: Data
			do {
				source = try Data(contentsOf: file)
			} catch {
				throw CatalogError.unreadable(tag, underlying: error)
			}

			// A translation that does not parse is a bug in this
			// repository rather than something to paper over at run
			// time, so these are fatal where a *formatting* error
			// later on is merely shown.
			guard let parseErrors = FluentErrors() else {
				throw CatalogError.outOfMemory
			}
			bundle.add(resource: source, collecting: parseErrors)
			if parseErrors.count > 0 {
				throw CatalogError.malformed(tag, problems: parseErrors.messages)
			}

			translations.append(Translation(
				tag: bundle.locale,
				// A translation that has not got round to
				// `language-name` yet still belongs in the menu,
				// under the only name we have for it.
				endonym: bundle.format("language-name") ?? bundle.locale,
				bundle: bundle))
		}
	}

	// MARK: - Negotiation

	/// Whether two canonical tags name the same language.
	///
	/// A tag out of the library is canonical -- language lowercase, script
	/// title case, region uppercase -- so the language is everything before
	/// the first hyphen and a plain comparison is enough. No locale library
	/// is needed for this, which is the point: what comes back is a string.
	private static func language(of tag: String) -> Substring {
		tag.prefix { $0 != "-" }
	}

	/// The script subtag, or `nil` when the tag does not state one. A
	/// four-letter subtag in that position is a script; a two-letter or
	/// three-digit one is a region, and there is then no script.
	private static func script(of tag: String) -> Substring? {
		let parts = tag.split(separator: "-")
		guard parts.count > 1, parts[1].count == 4 else { return nil }
		return parts[1]
	}

	private static func matches(_ have: String, _ want: String, pass: Int) -> Bool {
		guard language(of: have) == language(of: want) else { return false }
		switch pass {
		case 0:
			return have == want
		case 1:
			guard let left = script(of: have), let right = script(of: want) else {
				return true // unstated counts as agreeing
			}
			return left == right
		default:
			return true
		}
	}

	/// The translation that best serves the locales asked for.
	///
	/// Each requested locale is tried in turn, and for each one the
	/// translations are searched for the closest match: the same tag, then
	/// the same language and script, then merely the same language. Asking
	/// in that order matters -- somebody who asks for `fr` before `en`
	/// should get French even though an `en-US` translation exists and is a
	/// better match for nothing they said.
	///
	/// When nothing matches, the first entry: showing English is a better
	/// outcome than showing message names.
	public func negotiate(_ requested: [String]) -> Int {
		for want in requested {
			for pass in 0..<3 {
				for (index, translation) in translations.enumerated()
				where Self.matches(translation.tag, want, pass: pass) {
					return index
				}
			}
		}
		return 0
	}

	/// The same, for what the system says the user wants -- or for a
	/// colon-separated list, as `LANGUAGE` is written, when one is given.
	public func preferred(list: String? = nil) -> Int {
		negotiate(list.map { Fluent.locales(inList: $0) } ?? Fluent.preferredLocales())
	}

	public func applySystemFormats(to index: Int) {
		guard translations.indices.contains(index) else { return }
		translations[index].bundle.applySystemFormats()
	}

	// MARK: - Saying it

	public func format(_ id: String, in index: Int, args: FluentArgs? = nil) -> String? {
		guard translations.indices.contains(index) else { return nil }
		return translations[index].bundle.format(id, args: args, collecting: errors)
	}

	public func format(
		_ id: String, attribute: String, in index: Int, args: FluentArgs? = nil
	) -> String? {
		guard translations.indices.contains(index) else { return nil }
		return translations[index].bundle.format(
			id, attribute: attribute, args: args, collecting: errors)
	}

	// MARK: - Finding the translations

	/// Where the `.ftl` files are.
	///
	/// Inside the application bundle when there is one, which is what `make
	/// app` builds; otherwise `FLUENT_LOCALE_DIR`, and otherwise the
	/// repository's own copy relative to the working directory, which is
	/// where `swift run` leaves it.
	public static func defaultDirectory() -> URL {
		let manager = FileManager.default

		if let resources = Bundle.main.resourceURL {
			let inside = resources.appendingPathComponent("locales")
			if manager.fileExists(atPath: inside.path) { return inside }
		}

		if let named = ProcessInfo.processInfo.environment["FLUENT_LOCALE_DIR"] {
			return URL(fileURLWithPath: named)
		}

		return URL(fileURLWithPath: "../locales")
	}
}

File: examples/common/catalog.c
/* SPDX-FileCopyrightText: © 2026 Jeffrey C. Ollie <jeff@ocjtech.us> */
/* SPDX-License-Identifier: MIT */

#include "catalog.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/*
 * The translations shipped with the program.
 *
 * The first entry is the source locale and is the last resort, which is why
 * its translation must be complete.
 */
static const char *const shipped[] = { "en-US", "de", "fr", "fi", "pl", "ja" };
#define SHIPPED_LEN (sizeof(shipped) / sizeof(shipped[0]))

struct catalog {
	fluent_bundle *bundles[SHIPPED_LEN];
	/* The name each translation gives its own language, formatted once at
	 * load time because a menu asks for it on every redraw. */
	char *endonyms[SHIPPED_LEN];
	/* One list, cleared by the caller between screenfuls. Held here rather
	 * than made per call so that a window can format a dozen messages and
	 * then ask once what went wrong. */
	fluent_errors *errors;
};

/* -- reading a translation off disk --------------------------------------- */

/*
 * The whole of `dir/<tag>.ftl`, or NULL with a complaint on stderr.
 *
 * `*len` is set to the length in bytes, because that is what the library
 * wants: an `.ftl` file is text and not a C string, and nothing says it may
 * not contain a NUL.
 */
static char *read_locale(const char *dir, const char *tag, size_t *len)
{
	char path[512];
	int written = snprintf(path, sizeof(path), "%s/%s.ftl", dir, tag);
	if (written < 0 || (size_t)written >= sizeof(path)) {
		fprintf(stderr, "path too long for %s\n", tag);
		return NULL;
	}

	FILE *file = fopen(path, "rb");
	if (file == NULL) {
		perror(path);
		return NULL;
	}

	char *source = NULL;
	if (fseek(file, 0, SEEK_END) != 0) goto fail;

	long size = ftell(file);
	if (size < 0) goto fail;
	if (fseek(file, 0, SEEK_SET) != 0) goto fail;

	source = malloc((size_t)size + 1);
	if (source == NULL) goto fail;

	if (fread(source, 1, (size_t)size, file) != (size_t)size) goto fail;
	source[size] = '\0';

	fclose(file);
	*len = (size_t)size;
	return source;

fail:
	perror(path);
	free(source);
	fclose(file);
	return NULL;
}

/* -- loading -------------------------------------------------------------- */

catalog *catalog_open(const char *dir, bool isolating)
{
	catalog *cat = calloc(1, sizeof(*cat));
	if (cat == NULL) {
		fprintf(stderr, "out of memory\n");
		return NULL;
	}

	cat->errors = fluent_errors_new();
	if (cat->errors == NULL) {
		fprintf(stderr, "out of memory\n");
		goto fail;
	}

	for (size_t i = 0; i < SHIPPED_LEN; i += 1) {
		cat->bundles[i] = fluent_bundle_new(shipped[i]);
		if (cat->bundles[i] == NULL) {
			fprintf(stderr, "no bundle for %s\n", shipped[i]);
			goto fail;
		}

		/* The caller's choice, and it is a question about what will draw
		 * the text rather than about what the text is -- see catalog.h. */
		fluent_bundle_set_use_isolating(cat->bundles[i], isolating);

		size_t len;
		char *source = read_locale(dir, shipped[i], &len);
		if (source == NULL) goto fail;

		fluent_errors *parse_errors = fluent_errors_new();
		bool added = fluent_bundle_add_resource(cat->bundles[i], source,
							len, parse_errors);
		/* The bundle copied what it needed, so this can go now. */
		free(source);

		/*
		 * A translation that does not parse is a bug in this
		 * repository, not something to paper over at run time -- so
		 * these are complained about and fatal, where a *formatting*
		 * error later on is merely shown.
		 */
		for (size_t e = 0;
		     parse_errors != NULL && e < fluent_errors_len(parse_errors);
		     e += 1) {
			char *message = fluent_errors_message(parse_errors, e);
			fprintf(stderr, "%s.ftl: %s\n", shipped[i],
				message == NULL ? "(out of memory)" : message);
			fluent_string_free(message);
			added = false;
		}
		fluent_errors_free(parse_errors);

		if (!added) goto fail;

		/* Formatted now, while nothing can have gone wrong yet, and
		 * kept: a language menu asks for it on every redraw. */
		cat->endonyms[i] = fluent_bundle_format(
			cat->bundles[i], "language-name", NULL, NULL);
	}

	return cat;

fail:
	catalog_close(cat);
	return NULL;
}

void catalog_close(catalog *cat)
{
	if (cat == NULL) return;
	for (size_t i = 0; i < SHIPPED_LEN; i += 1) {
		fluent_string_free(cat->endonyms[i]);
		fluent_bundle_free(cat->bundles[i]);
	}
	fluent_errors_free(cat->errors);
	free(cat);
}

/* -- what we have --------------------------------------------------------- */

size_t catalog_len(const catalog *cat)
{
	(void)cat;
	return SHIPPED_LEN;
}

const char *catalog_tag(const catalog *cat, size_t index)
{
	if (index >= SHIPPED_LEN) return "";
	return fluent_bundle_locale(cat->bundles[index]);
}

const char *catalog_endonym(const catalog *cat, size_t index)
{
	if (index >= SHIPPED_LEN) return "";
	/* A translation that has not got round to `language-name` yet still
	 * belongs in the menu, under the only name we have for it. */
	if (cat->endonyms[index] == NULL)
		return fluent_bundle_locale(cat->bundles[index]);
	return cat->endonyms[index];
}

/* -- negotiation ---------------------------------------------------------- */

/*
 * Whether two canonical tags name the same language.
 *
 * A tag out of `fluent_tag_parse` is canonical -- language lowercase, script
 * title case, region uppercase, hyphens throughout -- so the language is
 * everything up to the first hyphen and a plain comparison is enough. No
 * locale library is needed to do this, which is the point: what comes back is
 * a string, and the application is free to treat it as one.
 */
static size_t language_len(const char *tag)
{
	const char *hyphen = strchr(tag, '-');
	return hyphen == NULL ? strlen(tag) : (size_t)(hyphen - tag);
}

static bool same_language(const char *have, const char *want)
{
	size_t len = language_len(have);
	return len == language_len(want) && strncmp(have, want, len) == 0;
}

/* The script subtag, or NULL when the tag does not state one. */
static const char *script_of(const char *tag)
{
	const char *rest = strchr(tag, '-');
	if (rest == NULL) return NULL;
	rest += 1;

	/* A four-letter subtag in this position is a script; a two-letter or
	 * three-digit one is a region, and there is then no script. */
	size_t len = language_len(rest); /* the next subtag, by the same rule */
	return len == 4 ? rest : NULL;
}

/* Whether two tags name the same script, counting "unstated" as agreeing. */
static bool scripts_agree(const char *a, const char *b)
{
	const char *left = script_of(a);
	const char *right = script_of(b);
	if (left == NULL || right == NULL) return true;
	return strncmp(left, right, 4) == 0;
}

/* Whether `have` serves `want` at a given level of exactness. */
static bool matches(const char *have, const char *want, int pass)
{
	if (!same_language(have, want)) return false;
	switch (pass) {
	case 0: return strcmp(have, want) == 0;
	case 1: return scripts_agree(have, want);
	default: return true;
	}
}

size_t catalog_negotiate(const catalog *cat, const fluent_tag *requested,
			 size_t requested_len)
{
	for (size_t i = 0; i < requested_len; i += 1)
		for (int pass = 0; pass < 3; pass += 1)
			for (size_t b = 0; b < SHIPPED_LEN; b += 1)
				if (matches(catalog_tag(cat, b),
					    requested[i].text, pass))
					return b;
	return 0;
}

size_t catalog_preferred(const catalog *cat, const char *list)
{
	fluent_tag requested[CATALOG_MAX_REQUESTED];
	size_t len = list != NULL
			     ? fluent_locales_from_list(list, requested,
							CATALOG_MAX_REQUESTED)
			     : fluent_preferred_locales(requested,
							CATALOG_MAX_REQUESTED);
	return catalog_negotiate(cat, requested, len);
}

void catalog_apply_system_formats(catalog *cat, size_t index)
{
	if (index >= SHIPPED_LEN) return;
	fluent_categories categories;
	fluent_categories_current(&categories);
	fluent_bundle_apply_categories(cat->bundles[index], &categories);
}

/* -- saying it ------------------------------------------------------------ */

char *catalog_format(catalog *cat, size_t index, const char *id,
		     const fluent_args *args)
{
	if (index >= SHIPPED_LEN) return NULL;
	return fluent_bundle_format(cat->bundles[index], id, args, cat->errors);
}

char *catalog_format_attribute(catalog *cat, size_t index, const char *id,
			       const char *attribute, const fluent_args *args)
{
	if (index >= SHIPPED_LEN) return NULL;
	return fluent_bundle_format_attribute(cat->bundles[index], id, attribute,
					      args, cat->errors);
}

void catalog_clear_errors(catalog *cat)
{
	fluent_errors_clear(cat->errors);
}

size_t catalog_error_count(const catalog *cat)
{
	return fluent_errors_len(cat->errors);
}

char *catalog_error_message(const catalog *cat, size_t index)
{
	return fluent_errors_message(cat->errors, index);
}
